src/core/ext/transport/chaotic_good/tcp_frame_transport.cc

```diff
@@ -0,0 +1,262 @@
+          // The write failures will be caught in TrySeq and exit
+          // loop. Therefore, only need to return Continue() in the
+          // last lambda function.
+          return Continue();
+        });
+  });
+}
+
+auto TcpFrameTransport::ReadFrameBytes() {
+  return TrySeq(
+      control_endpoint_.ReadSlice(TcpFrameHeader::kFrameHeaderSize),
+      [this](Slice read_buffer) {
+        auto frame_header =
+            TcpFrameHeader::Parse(reinterpret_cast<const uint8_t*>(
+                GRPC_SLICE_START_PTR(read_buffer.c_slice())));
+        GRPC_TRACE_LOG(chaotic_good, INFO)
+            << "CHAOTIC_GOOD: ReadHeader from:"
+            << ResolvedAddressToString(control_endpoint_.GetPeerAddress())
+                   .value_or("<<unknown peer address>>")
+            << " "
+            << (frame_header.ok() ? frame_header->ToString()
+                                  : frame_header.status().ToString());
+        return frame_header;
+      },
+      [this](TcpFrameHeader frame_header) {
+        return If(
+            // If the payload is on the connection frame
+            frame_header.payload_connection_id == 0,
+            // ... then read the data immediately and return an IncomingFrame
+            //     that contains the payload.
+            // We need to do this here so that we do not create head of line
+            // blocking issues reading later control frames (but waiting for a
+            // call to get scheduled time to read the payload).
+            [this, frame_header]() {
+              return Map(
+                  control_endpoint_.Read(frame_header.header.payload_length),
+                  [frame_header](absl::StatusOr<SliceBuffer> payload)
+                      -> absl::StatusOr<IncomingFrame> {
+                    if (!payload.ok()) return payload.status();
+                    return IncomingFrame(frame_header.header,
+                                         std::move(payload));
+                  });
+            },
+            // ... otherwise issue a read to the appropriate data endpoint,
+            //     which will return a read ticket - which can be used later
+            //     in the call promise to asynchronously wait for those bytes
+            //     to be available.
+            [this, frame_header]() -> absl::StatusOr<IncomingFrame> {
+              const auto padding =
+                  frame_header.Padding(options_.decode_alignment);
+              return IncomingFrame(
+                  frame_header.header,
+                  Map(data_endpoints_
+                          .Read(frame_header.payload_connection_id - 1,
+                                frame_header.header.payload_length + padding)
+                          .Await(),
+                      [padding](absl::StatusOr<SliceBuffer> payload)
+                          -> absl::StatusOr<SliceBuffer> {
+                        if (payload.ok()) payload->RemoveLastNBytes(padding);
```

Comment:
Please use RemoveLastNBytesNoInline to prevent creation of inlined slices

Reply:
done